cloudfour/cloudfour.com-patterns
# Modular scale value transforms for design tokens

## Overview

Adds modular scale value transforms for design tokens. This means you can specify certain types and they will automatically convert to a usable value based on our modular scale.

For example, these design tokens:

```yaml
props:
  - name: example_unitless
    value: 0
    type: modular
  - name: example_em
    value: 2
    type: modular/em
  - name: example_rem
    value: -1
    type: modular/rem
  - name: example_px
    value: 3
    type: modular/px
global:
  category: example 
```

Would result in the following values when used:

Sass Variable | Value
--- | ---
$example-unitless | 1
$example-em | 1.5625em
$example-rem | 0.8rem
$example-px | 31px

I didn't think this would be necessary until I started #490, which requires input sizes to coordinate with buttons. Since `height` is _still_ the only reliable way to do this, it became challenging to coordinate sizes that would align nicely with adjacent typography. Being able to specify shared dimensions via tokens without doing manual calculations is awesome.

In order to pull this off for both CSS and JavaScript, I centralized all of our Theo configuration into a single directory that exports the configured Theo object from its `index.js`. This has an additional benefit of moving the Sass importer out of `.storybook`, which was an issue I had with #569.

## Screenshots

![theo-ms](https://user-images.githubusercontent.com/69633/78086938-450b7200-7374-11ea-9926-17ecdfc8e01d.gif)

## Testing

1. Check out this branch locally.
1. Confirm that `npm run storybook` is still successful.
1. Open the `sizes.yml` token file.
1. Change the `control_pad_vertical` token value to a different step.
1. Confirm that the change takes effect for both the button component and the design token sizes documentation page.
1. Change the `type` of the token and confirm it still works. Valid types for padding are `modular/em`, `modular/rem` and `modular/px`.

---

- Resolves #549
